src/language-markdown/whitespace.js

```diff
@@ -0,0 +1,265 @@
+  if (
+    punctuationRegex.test(previousLastChar) ||
+    punctuationRegex.test(nextFirstChar)
+  ) {
+    return false;
+  }
+  // If the sentence uses the style that Space is injected in between CJ and alphanumerics, "\n" can be converted to Space.
+  return isInSentenceWithCJSpaces(path);
+}
+
+/**
+ * @param {WordKind | undefined} kind
+ * @returns {boolean} `true` if `kind` is CJK (including punctuation marks)
+ */
+function isCJK(kind) {
+  return (
+    kind === KIND_CJ_LETTER ||
+    kind === KIND_K_LETTER ||
+    kind === KIND_CJK_PUNCTUATION
+  );
+}
+
+/**
+ * @param {WordKind | undefined} kind
+ * @returns {boolean} `true` if `kind` is letter (not CJK punctuation)
+ */
+function isLetter(kind) {
+  return (
+    kind === KIND_NON_CJK || kind === KIND_CJ_LETTER || kind === KIND_K_LETTER
+  );
+}
+
+/**
+ * @param {WordKind | undefined} kind
+ * @returns {boolean} `true` if `kind` is western or Korean letters (divides words by Space)
+ */
+function isWesternOrKoreanLetter(kind) {
+  return kind === KIND_NON_CJK || kind === KIND_K_LETTER;
+}
+
+/**
+ * Returns whether “whitespace” (`"" | " " | "\n"`; see `WhitespaceValue`) can converted to `"\n"`
+ *
+ * @param {AstPath} path
+ * @param {WhitespaceValue} value
+ * @param {*} options
+ * @param {AdjacentNodes | undefined} [adjacentNodes]
+ * @returns {boolean} `true` if “whitespace” can be converted to `"\n"`
+ */
+function isBreakable(path, value, options, adjacentNodes) {
+  if (options.proseWrap !== "always") {
+    return false;
+  }
+  if (getAncestorNode(path, SINGLE_LINE_NODE_TYPES)) {
+    return false;
+  }
+  if (adjacentNodes === undefined) {
+    return true;
+  }
+  // Space & newline are always breakable
+  if (value !== "") {
+    return true;
+  }
+  // Simulates latin words; see #6516 (https://github.com/prettier/prettier/issues/6516)
+  // [Latin][""][Hangul] & vice versa => Don't break
+  // [han & kana][""][hangul], either
+  if (
+    (adjacentNodes.previous?.kind === KIND_K_LETTER &&
+      isLetter(adjacentNodes.next?.kind)) ||
+    (adjacentNodes.next?.kind === KIND_K_LETTER &&
+      isLetter(adjacentNodes.previous?.kind))
+  ) {
+    return false;
+  }
+  // https://en.wikipedia.org/wiki/Line_breaking_rules_in_East_Asian_languages
+  const isBreakingCJKLineBreakingRule =
+    (adjacentNodes.next?.value !== undefined &&
+      noBreakBefore.has(adjacentNodes.next?.value?.[0])) ||
+    (adjacentNodes.previous?.value !== undefined &&
+      noBreakAfter.has(adjacentNodes.previous?.value?.at(-1)));
+  // For "" (CJK and some non-space) higher priority than the following rule
+  if (isBreakingCJKLineBreakingRule) {
+    return false;
+  }
+  return true;
+}
+
+export {
+  isInSentenceWithCJSpaces,
+  canBeConvertedToSpace,
+  isCJK,
+  isLetter,
+  isWesternOrKoreanLetter,
+  isBreakable,
+  noBreakAfter,
+  noBreakBefore,
+  lineBreakBetweenTheseAndCJKConvertToSpaceSymbolSet
```

Comment:
Don't export things that are local to this file.

Reply:
Will remove other than `canBeConvertedToSpace, isBreakable`

Reply:
They're both implementation details of the `printLine` function. Please rename `printLine` to `printWhitespace`, move it to `whitespace.js` and rename `whitespace.js` to `print-whitespace.js`.

Reply:
@thorn0 As you say, `printLine` is not good name. I will fix it.

Reply:
Please rename `whitespace.js` to `print-whitespace.js`.